### ISSUE TITLE
extractClientIP always returns unknown client address if Http().bind(...) is used

### ISSUE THREAD

**Issue by akara** [akka/akka-http#344]
With around Akka 2.4.9, the connection ip propagation to the flow has moved into `Http().bindAndHandle(...)`. Also the code seems to force aggressive fusing, a good optimization. This leaves code using `Http().bind(...)` exposed. Of course, we can replicate the code from inside `bindAndHandle(...)`, but that's not a good practice and the use of `bind` may not get the benefits of `bindAndHandle`.

So why use `bind`? First it is not deprecated. Current code uses it. If it is deprecated, that should be documented with a migration guide. Second, there are certain things `bindAndHandle` cannot do that `bind` can do. If your request handling `Flow` needs to be mutable (i.e. changes based on runtime registration), `bindAndHandle` could not possibly do that. The signature takes a `Flow[HttpRequest, HttpResponse, Any]` as an input. That means the flow has to be constructed before the `bindAndHandle` happens and cannot be changed afterwards. Alternatively, if bindAndHandle would take `=> Flow[HttpRequest, HttpResponse, Any]` or `() => Flow[HttpRequest, HttpResponse, Any]`, and honor that behavior for every new connection created, it would have solved the issue. Perhaps a wish list item. Since we do not expect the `Flow` to change too often, some optimization/caching can still be honored and useful, and with the optimizations in `bindAndHandle`, it may be beneficial over having users call `bind` directly.

Given that we cannot avoid `bind` for now, here is code illustrating the issue:

``` scala
import akka.actor.ActorSystem
import akka.agent.Agent
import akka.http.scaladsl.Http
import akka.http.scaladsl.model._
import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.Route
import akka.stream._
import akka.stream.scaladsl.{Flow, Sink}
import com.typesafe.config.ConfigFactory

import scala.language.postfixOps

object DynamicHttpTest extends App {

  val config = ConfigFactory.parseString(
    """
      |akka.http.server.remote-address-header = on
    """.stripMargin)

  implicit val system = ActorSystem("my-system", config)
  implicit val executionContext = system.dispatcher
  implicit val materializer = ActorMaterializer()

  val initialRoute: Route =
    get {
      path("whoami") {
        extractClientIP { ip =>
          complete(ip.toString() + "\r\n")
        }
      } ~
      path("hello") {
        complete("Hello!\r\n")
      } ~
      path("swap") {
        swapToAlternate()
        complete("swapped\r\n")
      }
    }

  val alternateRoute: Route =
    get {
      path("whoami") {
        extractClientIP { ip =>
          complete(ip.toString() + "\r\n")
        }
      } ~
      path("hello") {
        complete("Hola amigo!\r\n")
      } ~
      path("swap") {
        swapToInitial()
        complete("swapped\r\n")
      }
    }

  val currentFlow = Agent[Flow[HttpRequest, HttpResponse, Any]](initialRoute)
  def swapToInitial() = currentFlow send (initialRoute: Flow[HttpRequest, HttpResponse, Any])
  def swapToAlternate() = currentFlow send (alternateRoute: Flow[HttpRequest, HttpResponse, Any])
  val port = 8080

  // bindAndHandle does not support dynamic flows inside.
  // The signature could have taken: => Flow[HttpRequest, HttpResponse, Any]
  // or: () => Flow[HttpRequest, HttpResponse]
  Http().bindAndHandle(currentFlow(), "0.0.0.0", port)

  // Now use bind alone on a different port.
  // The dynamic behavior based on the current value of the currentFlow agent is working properly here.
  // But now we lost the client ip extraction.
  val serverFlow = Http().bind("0.0.0.0", port + 1)
  serverFlow.to(Sink.foreach { conn =>
    conn.flow.join(currentFlow()).run()
  }).run()
}
```

To test this code, just build and run it (with the right dependencies, you know what they are). Then test it using some client that does not keep connections alive, like `curl`. Don't use a browser or the Akka client as that maintains a connection to a certain TTL. Here is the sample test:

``` bash
$ curl http://127.0.0.1:8080/whoami
127.0.0.1:62607
$ # This one is from bindAndHandle, works correctly
$ curl http://127.0.0.1:8081/whoami
unknown
$ # This one from bind (port 8081) and always returns unknown
$ curl http://127.0.0.1:8080/hello
Hello!
$ curl http://127.0.0.1:8080/swap
swapped
$ curl http://127.0.0.1:8080/hello
Hello!
$ # Of course, on port 8080 it is static, so your route is not swapped.
$ curl http://127.0.0.1:8081/hello
Hola amigo!
$ # Here it has swapped already for 8081 (bind) as the same agent is used.
$ curl http://127.0.0.1:8081/swap
swapped
$ curl http://127.0.0.1:8081/hello
Hello!
$ # And you can swap it back again to the original route. Viola!
```

Unfortunately, our code does depend on this behavior.

**Comment by jrudolph** [comment]
I wonder if this is not just a bug. Couldn't the same [bugfix](https://github.com/akka/akka-http/commit/a953c317f233ba09b6514cd35874da29a70677d6) as for `bindAndHandle` be applied to `bind` as well?

**Comment by jrudolph** [comment]
Thanks for the report, @akara.

**Comment by jrudolph** [comment]
Quickly put together a test show-casing the wrong behavior here: https://github.com/akka/akka-http/tree/w/344-support-remote-address-in-bind

**Comment by akara** [comment]
Thank you for being on it so quick @jrudolph! I'm just wondering with the current location of the above mentioned bugfix in `bindAndHandle` plus the optimizations being at that location, is it still worth providing a `bind` that solves the issue but may not be optimized? Performance numbers may come back biting you when people use the `bind` signature. Would it be better to provide a more dynamic version of `bindAndHandle` (name it what makes sense) that includes the optimizations with a signature like this:

```scala
  def bindAndHandle(
    interface: String, port: Int = DefaultPortForProtocol,
    connectionContext: ConnectionContext = defaultServerHttpContext,
    settings:          ServerSettings    = ServerSettings(system),
    log:               LoggingAdapter    = system.log)
  (handler: => Flow[HttpRequest, HttpResponse, Any])(implicit fm: Materializer):
  Future[ServerBinding]
```

Then deprecate or make the `bind` signature private. I don't see much use of the two layer stream produced by `bind` where you need to have the request flow inside the connection flow. Our only use is to allow for the shown dynamic behavior. But I might be totally off and there may be other creative uses of the connection stream - then you have to really think how to make sure that `bind` use case is optimized. Just a thought...

**Comment by jrudolph** [comment]
> I'm just wondering with the current location of the above mentioned bugfix in bindAndHandle plus the optimizations being at that location, is it still worth providing a bind that solves the issue but may not be optimized?

I think more likely we would add the fix just to `bind`. In the best case, `bindAndHandle` could then again be implemented in terms of `bind` without losing the extra optimization it does.

**Comment by rbudzko** [comment]
@jrudolph, I've removed majority of duplication and made sure both `bind` and `bindAndHandle` handles incoming remote addresses.

**Comment by jrudolph** [comment]
Fixed by #442.
